### ISSUE TITLE
sendmsg error 22

### ISSUE THREAD

**Issue by npd6** [npd6/npd6#27]
_From [npd6Proj...@gmail.com](https://code.google.com/u/107404320184732530601/) on October 04, 2011 15:34:13_

User reported via the blog site this:

I got a “Invalid argument” socket error at run time , is there anything wrong ?
Thanks.
—————————————————————————
[Nov 06 00:03:20] processNS: NS target addr: 200x:x030:x003:xxxx:020f:66ff:fe19:a516
[Nov 06 00:03:20] processNS: Local prefix: 200x:x030:x003:xxxx:0000:0000:0000:0000
[Nov 06 00:03:20] processNS: Neither white nor black listing in operation.
[Nov 06 00:03:20] addr6match: Called to match up to 64 bits.
[Nov 06 00:03:20] addr6match: Target and prefix matched up to bit position 64
[Nov 06 00:03:20] processNS: Target:prefix – Build NA response.
[Nov 06 00:03:20] processNS: Store target to list.
[Nov 06 00:03:20] storeTarget: New entry – recording.
[Nov 06 00:03:20] processNS: Outbound message built
[Nov 06 00:03:20] processNS: sendmsg returned with error 22 = Invalid argument
——–> got an error message from the log file .

_Original issue: http://code.google.com/p/npd6/issues/detail?id=27_

**Comment by npd6** [comment]
_From [npd6Proj...@gmail.com](https://code.google.com/u/107404320184732530601/) on October 04, 2011 06:35:07_

Asked user via blog comments to indicate distro, version, etc. Will then investigate.

**Comment by npd6** [comment]
_From [npd6Proj...@gmail.com](https://code.google.com/u/107404320184732530601/) on October 04, 2011 06:37:54_

Also requested via blog comment to run with --debug2 enabled.

**Comment by npd6** [comment]
_From [arthur77...@gmail.com](https://code.google.com/u/114887430629392048797/) on October 05, 2011 04:22:28_

User put more information:
Linux version 2.6.21.1,gcc version 3.4.6
Embedded platform with ARM11 CPU .
( I used ./npd6 -D -l /tmp/my.log on the previous log ,
  I think it has already with --debug2 enable 
)

**Comment by npd6** [comment]
_From [npd6Proj...@gmail.com](https://code.google.com/u/107404320184732530601/) on October 06, 2011 03:28:37_

Thanks for the update. Although off-hand I don't know what's causing this (!) I suspect that this being an ARM system has some sort of bearing on it - only tested on a variety if i386/amd64 "conventional" systems so far.

I'm keen to resolve this - will look at enhancing the debug around this area and then maybe you could run with that. Since I'm not immediately sure I can lay hands on some an ARM system to test with, I'd really appreciate if we can resolve this.

**Comment by npd6** [comment]
_From [npd6Proj...@gmail.com](https://code.google.com/u/107404320184732530601/) on October 06, 2011 05:12:03_

Did a quick test to eliminate anything relating to the oldish version of gcc. Was a long shot, but was the easiest one to eliminate first! So it's def a CPU/library issue. Hmmm.

**Comment by npd6** [comment]
_From [arthur77...@gmail.com](https://code.google.com/u/114887430629392048797/) on October 06, 2011 05:19:13_

no problem . I have the ARM platform on hand ,
if there's anything I can help , please just let me know .

**Comment by npd6** [comment]
_From [sgroa...@gmail.com](https://code.google.com/u/105896410811076206131/) on October 06, 2011 06:12:27_

Could you share the early/startup portion of your log with me? Up to about where this snippet ends would be great - in particular was keen to see the ifIndex debug line, as shown in this extract:

[Oct 06 14:08:58] readConfig: maxHops set to 255
[Oct 06 14:08:58] main: Using normalised prefix 2a01:0e35:8b25:7ea0:0000:0000:0000:0000/64
[Oct 06 14:08:58] main: ifIndex for eth0 is: 2
[Oct 06 14:08:58] open_packet_socket: Created PF_PACKET socket OK.
[Oct 06 14:08:58] open_packet_socket: packet socket bind to interface 2 OK
[Oct 06 14:08:58] open_packet_socket: setsockopt(SO_ATTACH_FILTER) OK
[Oct 06 14:08:58] init_sockets: open_packet_socket: OK.
[Oct 06 14:08:58] open_icmpv6_socket: setsockopt(IPV6_UNICAST_HOPS = 255) OK
[Oct 06 14:08:58] init_sockets: open_icmpv6_socket: OK.
[Oct 06 14:08:58] if_allmulti: Setting IFFALLMULTI if required.

To avoid posting in public, feel free to email anything to me via npd6project@gmail.com

Cheers.

**Comment by npd6** [comment]
_From [sgroa...@gmail.com](https://code.google.com/u/105896410811076206131/) on October 06, 2011 23:36:23_

Thanks for the log. I'll go through that and see what we see! I hope it shows something up. Will update a bit later.

**Comment by npd6** [comment]
_From [npd6Proj...@gmail.com](https://code.google.com/u/107404320184732530601/) on October 07, 2011 00:43:41_

OK - I see a few things in the log which I can look in to. Hopefully one of them will lead to the issue! Update soon.

**Comment by npd6** [comment]
_From [npd6Proj...@gmail.com](https://code.google.com/u/107404320184732530601/) on October 07, 2011 00:44:01_

**Status:** Started

**Comment by npd6** [comment]
_From [npd6Proj...@gmail.com](https://code.google.com/u/107404320184732530601/) on October 07, 2011 01:02:09_

Doesn't have latest trunk which has additional debug. But suspicious about the src and/or dst addrs in the NS received. Suspect that they are causing an issue. Need to check.

**Comment by npd6** [comment]
_From [npd6Proj...@gmail.com](https://code.google.com/u/107404320184732530601/) on October 07, 2011 01:09:45_

I think I see it. The NS comes in here and shows this:

[Nov 09 01:34:26] processNS: src addr = ::
[Nov 09 01:34:26] processNS: dst addr = ff02::<obfuscated......>

We'll be sending back to that source address. Which ain't gonna work. :-) need to see if NS should come in with null-source (it may well be legit - need to check the RFC) and if so decide what the hell we do about it!!!

Question to the bug reporter: are you aware why in your scenario the NS comes in with a '::' as the source?

**Comment by npd6** [comment]
_From [npd6Proj...@gmail.com](https://code.google.com/u/107404320184732530601/) on October 07, 2011 01:25:39_

Yes, that'll be it... The NS comes in with src = :: which is, per RFC5156 , unspecified.

I think (wouldn't swear to it...) that this should only be used as a source if the sending device has initialised and not yet got its own addr.

Will double check, but I think it's invalid. Hence we cannot reply. Now it's true we should spot this and log it better rather than falling off the sendmsg() !! But I'm not sure we can do anything about it other than trap it - we simply do not know where to send an NA response.

**Comment by npd6** [comment]
_From [npd6Proj...@gmail.com](https://code.google.com/u/107404320184732530601/) on October 07, 2011 01:46:15_

So reading RFC 4861 we find that a src addr of a NS must be: "Either an address assigned to the interface from which this message is sent or (if Duplicate Address Detection is in progress [ADDRCONF]) the unspecified address."

We seeing the unspec addr, so we assume DAD is in operation. But we don't and then fail. OK, clear bug.

Looking beyond though, what should we do apart from trap this special case? We're in RFC 4862, section 5.4.2. I need coffee before working through that one... :-)

But in summary: we can resolve this. Could the bug reporter tell me though whether this is in-line with their understanding too? Cheers.

**Comment by npd6** [comment]
_From [npd6Proj...@gmail.com](https://code.google.com/u/107404320184732530601/) on October 07, 2011 02:12:03_

Doh! Obvious now I look at is: the dst addr begins ff02::1:..... which is the allnodes multicast address. Yes, what we are seeing here is entirely legal (but unexpected!) DAD behaviour. Just need to decide what to do. Update soon.

**Comment by npd6** [comment]
_From [npd6Proj...@gmail.com](https://code.google.com/u/107404320184732530601/) on October 07, 2011 02:22:07_

OK, got it. RFC 4862 section 5.4.3. In totality it read thus:

<quote>
 On receipt of a valid Neighbor Solicitation message on an interface,
   node behavior depends on whether or not the target address is
   tentative.  If the target address is not tentative (i.e., it is
   assigned to the receiving interface), the solicitation is processed
   as described in [ RFC4861 ].  If the target address is tentative, and
   the source address is a unicast address, the solicitation's sender is
   performing address resolution on the target; the solicitation should
   be silently ignored.  Otherwise, processing takes place as described
   below.  In all cases, a node MUST NOT respond to a Neighbor
   Solicitation for a tentative address.

   If the source address of the Neighbor Solicitation is the unspecified
   address, the solicitation is from a node performing Duplicate Address
   Detection.  If the solicitation is from another node, the tentative
   address is a duplicate and should not be used (by either node).  If
   the solicitation is from the node itself (because the node loops back
   multicast packets), the solicitation does not indicate the presence
   of a duplicate address.
</quote>

So clearly what we should be doing is detecting these and ignoring them (with an appropriate log) DAD is entirely outside the scope of npd6, and should be handled by the existing mechanism in place. All we need to do is ignore.

Question to the bug reporter: given that, what should happen here is that the NS received (of this type) is silently ignored, and not replied to. Which is what I'll do. However please confirm that if you ignore the logging of these errors for the moment (since they are not fatal) that npd6 _does_ respond correctly for you upon receipt of a "normal" NS? In other words, this bug, while confusing, shouldn't actually affect the normal behaviour. Can you check this?

Meanwhile I'll implement a fix for detecting this condition.

**Comment by npd6** [comment]
_From [arthur77...@gmail.com](https://code.google.com/u/114887430629392048797/) on October 07, 2011 03:07:41_

Oops! I am not aware the programmer has posted so many comments during the past hours :-(

well , from the log ,
Nov 09 01:39:21] storeTarget: New entry - recording.
[Nov 09 01:39:21] processNS: Outbound message built
[Nov 09 01:39:21] processNS: sendmsg returned with error 22 = Invalid argument

my question is :If the new IPv6 client found ,but the sendmsg error ,
will the npd6 still add neigh proxy for this client ?

because , from my previous test , it's not only a log confusing ,
but I need to manually add linux comment , like "ip -6 neigh add proxy 200x:x030:x003:xxxx:dddd:aaaa:bbbb:cccc dev eth2" for the client to get the ipv6 packet from WAN.

**Comment by npd6** [comment]
_From [npd6Proj...@gmail.com](https://code.google.com/u/107404320184732530601/) on October 07, 2011 03:25:38_

Well, what we see here is not a normal Neighbor Solicitation. As you will know, the normal behaviour is that one device says:
- Are you address XXX?
- And device XXX says "Yes, I'm here!"

and then off we go. That's what npd6 facilitates, without the "ip -6 neigh add proxy" you mention.

But the neighbor solicitation I see in the log is not of that type at all. It's a special case which should not be replied to (indeed cannot be replied to even if we wanted!) It's a device which is in some sort of initialising state attempting to ensure that duplicate addresses are not in use.

So what npd6 should do is instead of gracelessly giving an error 22 on the sendmsg, it should silently ignore these (and if debug turned on log their presence)

However it should not do more than that. What it will do is if a "normal" neighbor solicitation comes in it will match and reply.

What slightly confuses me here is why your "other" device keeps repeatedly sending the "special DAD" NS without stopping. I'll try and work out why, even though I'm pretty sure it's outside the scope of npd6 handling.

Can you tell me this: in the log you sent me, the dst address on the NS that we see, the one beginning ff02::1..... There are several variations in the log (i.e. the final part varies) Who/what are those devices? I suspect they are just "other nodes", or, potentially, they could even be "this" mode looping them back.

**Comment by npd6** [comment]
_From [arthur77...@gmail.com](https://code.google.com/u/114887430629392048797/) on October 07, 2011 04:21:06_

Because I manually issue the linux command   "ip -6 neigh add proxy ....." for the clients , then have the client ping  ipv6.google.com , I think 
case 1:
The multicast packet ff02::1:nodeA ,  ff02::1:nodeB ,  ff02::1:nodeC, is the NS sent from the ARM11 box to client A,B C , (and then the clients will feedback NA to the ARM11 box , then the clients can get google's ping reply )  ,
case 2:
or because the clients just join ARM11 ? 
I think it's likely case 1.
( will confirm while I back to the lab)

**Comment by npd6** [comment]
_From [npd6Proj...@gmail.com](https://code.google.com/u/107404320184732530601/) on October 07, 2011 04:42:22_

Now we simply check for null src addr and log at DEBUG level:

"Unspecified src addr - DAD activity. Ignoring NS."

and then silently ignore the NS.

**Status:** Done_Pending

**Comment by npd6** [comment]
_From [npd6Proj...@gmail.com](https://code.google.com/u/107404320184732530601/) on October 07, 2011 04:46:34_

Committed in 81

**Comment by npd6** [comment]
_From [npd6Proj...@gmail.com](https://code.google.com/u/107404320184732530601/) on October 07, 2011 04:49:59_

Fixed in mainline trunk.

Will cut a new rev over the weekend or early next week - holding off pending any other fixes which need to be rolled in.

Could you test by grabbing the mainline using:
svn checkout http://npd6.googlecode.com/svn/trunk/ npd6-read-only

or even just use the web interface to pull the changed file ip6.c - it's a small change.

When tested you should now see no error=22 on the sendmsg(). With no debug enabled the DAD-NS will be silently ignored. If debug is enabled it will log a line for each, and then ignore.

**Comment by npd6** [comment]
_From [sgroa...@gmail.com](https://code.google.com/u/105896410811076206131/) on October 10, 2011 01:19:16_

Just cut rev 0.4.4 to incorporate this fix. Will keep bug open a little longer pending feedback from the reporter.

**Comment by npd6** [comment]
_From [sparkm...@gmail.com](https://code.google.com/u/112882192538038678901/) on October 17, 2011 22:54:51_

Ah, I see what you're missing.  You assume that replying to a packet sourced from :: is impossible, but you're failing to take into account the Layer 2 MAC address.

Here's a standard neighbor solicitation:

22:30:03.121425 IP6 2001:470:8103:0:2e0:4dff:fe39:9407 > ff02::1:ffdb:550d: ICMP6, neighbor solicitation, who has 2001:470:8103:0:21b:fcff:fedb:550d, length 32
22:30:03.121456 IP6 2001:470:8103:0:21b:fcff:fedb:550d > 2001:470:8103:0:2e0:4dff:fe39:9407: ICMP6, neighbor advertisement, tgt is 2001:470:8103:0:21b:fcff:fedb:550d, length 32

Notice that there's a single request, and a single response:
- X->*: where is Y?
- Y->X: here I am!

For comparison, here's an npd6 neighbor solicitation:

22:31:43.161473 IP6 2001:470:8103:0:2e0:4dff:fe39:9407 > ff02::1:ff00:3: ICMP6, neighbor solicitation, who has 2001:470:8103::f00:3, length 32
22:31:43.162543 IP6 2001:470:8103:0:21b:fcff:fedb:550d > ff02::1:ff39:9407: ICMP6, neighbor solicitation, who has 2001:470:8103:0:2e0:4dff:fe39:9407, length 32
22:31:43.162833 IP6 2001:470:8103:0:2e0:4dff:fe39:9407 > 2001:470:8103:0:21b:fcff:fedb:550d: ICMP6, neighbor advertisement, tgt is 2001:470:8103:0:2e0:4dff:fe39:9407, length 32
22:31:43.162843 IP6 2001:470:8103:0:21b:fcff:fedb:550d > 2001:470:8103:0:2e0:4dff:fe39:9407: ICMP6, neighbor advertisement, tgt is 2001:470:8103::f00:3, length 32

Notice that there are two additional requests:
- X->*: where is Y?
- Y->*: where is X?
- X->Y: here I am!
- Y->X: here I am!

The standard ND handshake is only 2-way because the response is sent directly to the requester's MAC address.

npd6 ignores the MAC address, and only sends the response to the IPv6 address.  Thus, an additional round trip is necessary to figure out who the sender is.  When the sender is ::, that step is impossible.

Aside from the loss of efficiency, ignoring Duplicate Address Detection is potentially harmful.  Basically, the npd6 node is claiming a large swath of IPv6 addresses for itself, without having the courtesy to tell competing nodes that those addresses are no longer available.

Long story short, in order to be RFC compliant, you'll need to grab the MAC address from incoming packets, and respond to MAC addresses in your outgoing packets.  I don't know precisely how to do this, but it's likely that an ICMP socket is too high-level for your outbound packets.

**Comment by npd6** [comment]
_From [sparkm...@gmail.com](https://code.google.com/u/112882192538038678901/) on October 17, 2011 22:59:51_

(And in case it wasn't obvious, my tcpdump output doesn't actually include the MAC addresses.  They show up in "tcpdump -v", but I didn't want to paste that much text.)

**Comment by npd6** [comment]
_From [sparkm...@gmail.com](https://code.google.com/u/112882192538038678901/) on October 17, 2011 23:13:59_

Actually, in order to be _truly_ RFC compliant, the npd6 node would have to send out billions of DAD messages to claim all of its addresses.  That's obviously not going to work, so there's a line to walk between compliance and simplicity...

In any case, reducing the handshake from 4 packets to 2 would be a nice goal.

**Comment by npd6** [comment]
_From [sgroa...@gmail.com](https://code.google.com/u/105896410811076206131/) on January 26, 2012 23:54:53_

Been starting a new job this month... So kinda neglected the project while I settle. Stressful :-)

Will be catching up with this and other issues real soon now.

Thanks for your patience!

**Comment by npd6** [comment]
_From [npd6Proj...@gmail.com](https://code.google.com/u/107404320184732530601/) on February 28, 2012 08:07:38_

Massive backlog, getting cleared. I'm genuinely pleased to get discussion on this point, but I am very perplexed also. Splitting the feedback from Comment 24 into 2 chunks, and taking one at a time...

The example tcpdump shown for a "npd6 neighbor solicitation". We seem to have things the wrong way around! npd6 _listens_ for N-Solicitations and (potentially - depending upon config) itself replies with N-Advertisements.

In the example we see 2 _incoming_ NS, from an adjacent system. They are not produced by npd6. It may or may not respond to them, based upon config. The 2 NAs seen going out could well have come from npd6 - if so configured. So there's really no "2 way" versus "4 way" debate at all. 

Also: "npd6 ignores the MAC address, and only sends the response to the IPv6 address." I may have misinterpreted what you mean there. But as it stands, that doesn't make any practical sense. :-) It is, pedantically, accurate! But the behaviour is also correct. npd6 does not, nor should it, care about MAC addresses!

If in your environment you are seeing "doubled up" NSs, then look outside the system with npd6 to ascertain why the other system is sending them.

**Comment by npd6** [comment]
_From [npd6Proj...@gmail.com](https://code.google.com/u/107404320184732530601/) on February 28, 2012 08:21:27_

Feeedback re. comment 24:

"Basically, the npd6 node is claiming a large swath of IPv6 addresses for itself,..."

Not sure about "claiming", but advertising reach-ability of a (configurable) swathe of addresses - yes. That is indeed its purpose!  As per previous, npd6 does not advertise anything unsolicited. It responds to queries.

Re. DAD: again, I think this is the wrong way around. It is entirely correct that npd6 should ignore DAD-related messages. And it has no need to send any of its own.

Wanted to clarify in case anyone reading gets confused.

**Comment by npd6** [comment]
_From [npd6Proj...@gmail.com](https://code.google.com/u/107404320184732530601/) on February 28, 2012 08:23:35_

Closing.

**Status:** Done

**Comment by npd6** [comment]
_From [sparkm...@gmail.com](https://code.google.com/u/112882192538038678901/) on February 28, 2012 09:51:33_

Responding to #28.
I will explain the 4-way handshake in more detail:

X is a machine elsewhere on the network.
Y is the machine running npd6.

1) X->_: where is Y? (sent from the other machine)
2) Y->_: where is X? (generated by the npd6 machine's kernel, because npd6
                      is trying to send a response packet to X's IP address,
                      without specifying X's MAC address.  The actual npd6
                      packet is still buffered in the kernel at this point.)
3) X->Y: here I am! (sent from the other machine)
4) Y->X: here I am! (finally, this is the response from npd6)

Steps 2 and 3 are redundant, and exist because npd6 is ignoring Layer 2.

"But the behaviour is also correct. npd6 does not, nor should it, care about MAC addresses!"

I reiterate that this is _not_ correct.  The neighbor discovery protocol is all about mapping between Layer 2 (MAC) and Layer 3 (IPv6) addresses.  The packets involved may contain IPv6 addresses, but they are fundamentally delivered from MAC to MAC.

**Comment by npd6** [comment]
_From [sgroa...@gmail.com](https://code.google.com/u/105896410811076206131/) on February 28, 2012 12:39:53_

OK - I think I see what you're seeing well, can picture it!)

Yes, _depending upon the specific network setup_ there may well be a legitimate (in fact required) "doubling up". If depends upon the players. As per RFC 4861, para 7.2 (RFCs... yawn,... can't live with 'em, can't, etc.) goes out of its way to state:

<quote>
If a Neighbor Cache entry does not
   exist for the source of such a message, Address Resolution will be
   required before unicast communications with that address can begin.
   This is particularly relevant for unicast responses to solicitations
   where an additional packet exchange is required for advertisement
   delivery.
</quote>

Which I guess is what you see? In the setups where I've bothered to look, I've not seen that yet. But as per, it's entirely OK and implicitly required.

So steps 2 & 3 should not be seen as redundant - if they were not required, they would not be seen. Bear in mind that the npd6 does not itself generate the second NS (directly). You also agree on this.

In para 7.2.4 it defines some paths re use of (optional) received link-layer address in the corresponding NA. I will re-read it in the morning with a fresh mind but I don't think we go outside of that.

So in summary:
- what you see seems valid in RFC terms.
- not everyone will see it.
- in pure RFC legalese, I _think_ we're compliant. But if you can point at a specific issue with that whole 7.2 section in the RFC then I'm all ears!! 

If when you say "...because npd6 is ignoring Layer 2." you can point at the RFC non-compliance, I'll sincerely be happy. I'm not trying to be a smart-ass about it. :-) 

My worry is that I've often seen situations (in general) in the past where doing something differently would be better! But may not stick to the RFC. That's what is on my mind here.

# 

PS And any kind reader who wishes to point out that npd6 itself violates "architectural" RFCs by its very functionality... Well, you have a point, albeit entirely moot due to the non-subnettable IPv6 implementations which are popping up all over the place. Let's not go there in a bug report... ;-)
